k8s/README.md

```diff
@@ -0,0 +1,313 @@
+rk-2      0/1       Terminating   0         9m
+rk-2      0/1       Terminating   0         9m
+rk-2      0/1       Terminating   0         9m
+rk-1      0/1       Terminating   0         10m
+rk-1      0/1       Terminating   0         10m
+rk-1      0/1       Terminating   0         10m
+rk-0      0/1       Terminating   0         11m
+rk-0      0/1       Terminating   0         11m
+rk-0      0/1       Terminating   0         11m
+```
+
+Reapply the manifest in `raftkeeper-k8s-deployment.yaml`.
+
+```shell
+kubectl apply -f raftkeeper-k8s-deployment.yaml
+```
+
+This creates the `rk` StatefulSet object, but the other API objects in the manifest are not modified because they already exist.
+
+Watch the StatefulSet controller recreate the StatefulSet's Pods.
+
+```shell
+kubectl get pods -w -l app=rk
+```
+
+Once all three Pods are Running and Ready, use `CTRL-C` to terminate kubectl.
+
+```
+NAME      READY     STATUS    RESTARTS   AGE
+rk-0      0/1       Pending   0          0s
+rk-0      0/1       Pending   0         0s
+rk-0      0/1       ContainerCreating   0         0s
+rk-0      0/1       Running   0         19s
+rk-0      1/1       Running   0         40s
+rk-1      0/1       Pending   0         0s
+rk-1      0/1       Pending   0         0s
+rk-1      0/1       ContainerCreating   0         0s
+rk-1      0/1       Running   0         18s
+rk-1      1/1       Running   0         40s
+rk-2      0/1       Pending   0         0s
+rk-2      0/1       Pending   0         0s
+rk-2      0/1       ContainerCreating   0         0s
+rk-2      0/1       Running   0         19s
+rk-2      1/1       Running   0         40s
+```
+
+Use the command below to get the value you entered during the [sanity test](https://kubernetes.io/docs/tutorials/stateful-application/zookeeper/#sanity-testing-the-ensemble), from the `rk-2` Pod.
+
+```shell
+kubectl exec rk-2 zkCli.sh -server localhost:8101 get /hello
+```
+
+Even though you terminated and recreated all of the Pods in the `rk` StatefulSet, the ensemble still serves the original value.
+
+```
+WATCHER::
+
+WatchedEvent state:SyncConnected type:None path:null
+world
+```
+
+The `volumeClaimTemplates` field of the `rk` StatefulSet's `spec` specifies a PersistentVolume provisioned for each Pod. The `StatefulSet` controller generates a `PersistentVolumeClaim` for each Pod in the `StatefulSet`.
+
+Use the following command to get the `StatefulSet`'s `PersistentVolumeClaims`.
+
+```shell
+kubectl get pvc -l app=rk
+```
+
+## Testing for liveness
+
+The Pod `template` for the `rk` `StatefulSet` specifies a liveness probe.
+
+```yaml
+livenessProbe:
+  tcpSocket:
+    port: 8101
+  initialDelaySeconds: 30
+  periodSeconds: 10
+  timeoutSeconds: 5
+```
+
+The probe tries to connect to 8101 port to test the server's health.
+
+When the liveness probe for the ZooKeeper process fails, Kubernetes will automatically restart the process for you, ensuring that unhealthy processes in the ensemble are restarted.
```

Comment:
Where did you set up the restart function?

Reply:
K8s will restart the container for us.